website/docs/configuration.md
---
title: Configuration
---

## Configuring the server

Wave allows starting Wave server in 2 ways:

- via `wave run` command - this automatically starts Wave server (waved) under the hood, useful for development.
- via Wave server binary (waved) - useful during deployment or when you need to run your Wave server on a different machine than your app.

Wave can be configured via configuration (`.env`) file, environment variables or command line arguments with the following priority: `cmd arg > env var > config > default`.

<!-- CREDIT: https://www.tablesgenerator.com/markdown_tables. -->
<!-- https://github.com/h2oai/wave/issues/2256 -->

| ENV var or config (wave run or waved)  | CLI args (waved)                      | Description                                                                                                                                                                                                                                                                                                          |
| -------------------------------------- | ------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| H2O_WAVE_ACCESS_KEY_ID                 | -access-key-id string                 | default API access key ID (default "access_key_id")                                                                                                                                                                                                                                                                  |
| H2O_WAVE_ACCESS_KEY_SECRET             | -access-key-secret string             | default API access key secret (default "access_key_secret")                                                                                                                                                                                                                                                          |
| H2O_WAVE_ACCESS_KEYCHAIN               | -access-keychain string               | path to file containing API access keys (default ".wave-keychain")                                                                                                                                                                                                                                                   |
|                                        | -compact string                       | compact AOF log                                                                                                                                                                                                                                                                                                      |
|                                        | -create-access-key                    | generate and add a new API access key ID and secret pair to the keychain                                                                                                                                                                                                                                             |
| H2O_WAVE_DATA_DIR                      | -data-dir string                      | directory to store site data (default "./data").                                                                                                                                                                                                                                                                     |
| H2O_WAVE_DEBUG [^1]                    | -debug                                | enable debug mode (profiling, inspection, etc.)                                                                                                                                                                                                                                                                      |
| H2O_WAVE_EDITABLE [^1]                 | -editable                             | allow users to edit web pages                                                                                                                                                                                                                                                                                        |
| H2O_WAVE_FORWARDED_HTTP_HEADERS        | -forwarded-http-headers string        | comma-separated list of case-insensitive HTTP header keys to forward to the Wave app from the browser WS connection. If not specified, defaults to '\*' - all headers are allowed. If set to an empty string, no headers are forwarded.                                                                              |
| H2O_WAVE_HTTP_HEADERS_FILE             | -http-headers-file string             | path to a MIME-formatted file containing additional HTTP headers to add to responses from the server                                                                                                                                                                                                                 |
| H2O_WAVE_INIT                          | -init string                          | initialize site content from AOF log                                                                                                                                                                                                                                                                                 |
| H2O_WAVE_LISTEN                        | -listen string                        | listen on this address (default ":10101")                                                                                                                                                                                                                                                                            |
|                                        | -list-access-keys                     | list all the access key IDs in the keychain                                                                                                                                                                                                                                                                          |
| H2O_WAVE_BASE_URL                      | -base-url string                      | the base URL (path prefix) to be used for resolving relative URLs (e.g. /foo/ or /foo/bar/, without the host (default "/")                                                                                                                                                                                           |
| H2O_WAVE_MAX_CACHE_REQUEST_SIZE        | -max-cache-request-size string        | maximum allowed size of HTTP requests to the server cache (e.g. 5M or 5MB or 5MiB) (default "5M")                                                                                                                                                                                                                    |
| H2O_WAVE_MAX_PROXY_REQUEST_SIZE        | -max-proxy-request-size string        | maximum allowed size of proxied HTTP requests (e.g. 5M or 5MB or 5MiB) (default "5M")                                                                                                                                                                                                                                |
| H2O_WAVE_MAX_PROXY_RESPONSE_SIZE       | -max-proxy-response-size string       | maximum allowed size of proxied HTTP responses (e.g. 5M or 5MB or 5MiB) (default "5M")                                                                                                                                                                                                                               |
| H2O_WAVE_MAX_REQUEST_SIZE              | -max-request-size string              | maximum allowed size of HTTP requests to the server (e.g. 5M or 5MB or 5MiB) (default "5M")                                                                                                                                                                                                                          |
| H2O_WAVE_NO_STORE [^1]                 | -no-store                             | disable storage (scripts and multicast/broadcast apps will not work)                                                                                                                                                                                                                                                 |
| H2O_WAVE_NO_LOG [^1]                   | -no-log                               | disable AOF logging (connect/disconnect and diagnostic logging messages are not disabled)                                                                                                                                                                                                                            |
| H2O_WAVE_OIDC_AUTH_URL_PARAMS          | -oidc-auth-url-params string          | additional URL parameters to pass during OIDC authorization, in the format "key:value", comma-separated, e.g. "foo:bar,qux:42"                                                                                                                                                                                       |
| H2O_WAVE_OIDC_CLIENT_ID                | -oidc-client-id string                | OIDC client ID                                                                                                                                                                                                                                                                                                       |
| H2O_WAVE_OIDC_CLIENT_SECRET            | -oidc-client-secret string            | OIDC client secret                                                                                                                                                                                                                                                                                                   |
| H2O_WAVE_OIDC_END_SESSION_URL          | -oidc-end-session-url string          | OIDC end session URL                                                                                                                                                                                                                                                                                                 |
| H2O_WAVE_OIDC_PROVIDER_URL             | -oidc-provider-url string             | OIDC provider URL                                                                                                                                                                                                                                                                                                    |
| H2O_WAVE_OIDC_REDIRECT_URL             | -oidc-redirect-url string             | OIDC redirect URL                                                                                                                                                                                                                                                                                                    |
| H2O_WAVE_OIDC_POST_LOGOUT_REDIRECT_URL | -oidc-post-logout-redirect-url string | OIDC post logout redirect URL                                                                                                                                                                                                                                                                                        |
| H2O_WAVE_OIDC_SCOPES                   | -oidc-scopes                          | OIDC scopes separated by comma (default "openid,profile")                                                                                                                                                                                                                                                            |
| H2O_WAVE_OIDC_SKIP_LOGIN [^1]          | -oidc-skip-login                      | don't show the built -in login form during OIDC authorization                                                                                                                                                                                                                                                        |
| H2O_WAVE_PRIVATE_DIR [^2]              | -private-dir value                    | additional directory to serve files from (authenticated users only), in the format "[url-path]@[filesystem-path]", e.g. "/public/files/@/some/local/path" will host /some/local/path/foo.txt at /public/files/foo.txt; multiple directory mappings allowed; paths need to be relative to Wave server binary location |
| H2O_WAVE_PUBLIC_DIR [^2]               | -public-dir value                     | additional directory to serve files from, in the format "[url-path]@[filesystem-path]", e.g. "/public/files/@/some/local/path" will host /some/local/path/foo.txt at /public/files/foo.txt; multiple directory mappings allowed; paths need to be relative to Wave server binary location                            |
| H2O_WAVE_PROXY [^1]                    | -proxy                                | enable HTTP proxy (for IDE / language server support only - not recommended for internet-facing websites)                                                                                                                                                                                                            |
|                                        | -remove-access-key string             | remove the specified API access key ID from the keychain                                                                                                                                                                                                                                                             |
| H2O_WAVE_SESSION_EXPIRY                | -session-expiry string                | session cookie lifetime duration (e.g. 1800s or 30m or 0.5h) (default "720h")                                                                                                                                                                                                                                        |
| H2O_WAVE_SESSION_INACTIVITY_TIMEOUT    | -session-inactivity-timeout string    | session inactivity timeout duration (e.g. 1800s or 30m or 0.5h) (default "30m")                                                                                                                                                                                                                                      |
| H2O_WAVE_TLS_CERT_FILE                 | -tls-cert-file string                 | path to certificate file (TLS only)                                                                                                                                                                                                                                                                                  |
| H2O_WAVE_TLS_KEY_FILE                  | -tls-key-file string                  | path to private key file (TLS only)                                                                                                                                                                                                                                                                                  |
| H2O_WAVE_NO_TLS_VERIFY [^1]            | -no-tls-verify                        | do not verify TLS certificates during external communication - DO NOT USE IN PRODUCTION                                                                                                                                                                                                                              |
|                                        | -version                              | print version and exit                                                                                                                                                                                                                                                                                               |
| H2O_WAVE_KEEP_APP_LIVE [^1]            | -keep-app-live                        | do not unregister unresponsive apps (default false)                                                                                                                                                                                                                                                                  |
| H2O_WAVE_WEB_DIR                       | -web-dir string                       | directory to serve web assets from (default "./www")                                                                                                                                                                                                                                                                 |
| H2O_WAVE_CONF                          | -conf string                          | path to a configuration file (default ".env")                                                                                                                                                                                                                                                                        |
| H2O_WAVE_PING_INTERVAL                 | -ping-interval string                 | how often should ping messages be sent (e.g. 60s or 1m or 0.1h) to keep the websocket connection alive (default "50s")                                                                                                                                                                                               |
| H2O_WAVE_RECONNECT_TIMEOUT             | -reconnect-timeout string             | Time to wait for reconnect before dropping the client (default "2s")                                                                                                                                                                                                                                                 |
| H2O_WAVE_ALLOWED_ORIGINS               | -allowed-origins string               | comma-separated list of allowed origins (e.g. http://foo.com) for websocket upgrades, use `*` to allow all origins                                                                                                                                                                                                   |

[^1]: `1`, `t`, `true` to enable; `0`, `f`, `false` to disable (case insensitive).

[^2]: Use OS-specific path list separator to specify multiple arguments - `:` for Linux/OSX and `;` for Windows. For example, `H2O_WAVE_PUBLIC_DIR=/images/@./files/images:/downloads/@./files/downloads`.

### File paths

All the configuration options that expect a path as value (public/private dir, data dir etc.) need the path provided to be either absolute or relative to a directory where `waved` is stored.

Those that do not start `waved` manually (but use `wave run` instead) and would like to use relative paths need to know that `waved` is stored in their 'virtualenv' (venv) folder (when installed via `pip`). If the `venv` folder is within the project root, then all the paths relative to your project dir need to be prepended with `../` (up one directory).

### Supported size units (case insensitive)

- Exabyte: `E` / `EB` / `EIB`.
- Petabyte: `P` / `PB` / `PIB`.
- Terabyte: `T` / `TB` / `TIB`.
- Gigabyte: `G` / `GB` / `GIB`.
- Megabyte: `M` / `MB` / `MIB`.
- Kilobyte: `K` / `KB` / `KIB`.
- Byte: `B`

### Suported time units

- Nanosecond: `ns`.
- Microsecond: `us` (or `µs`).
- Millisecond `ms`.
- Second: `s`.
- Minute: `m`.
- Hour: `h`.

### Public/Private dirs

When the wave server (waved) is started as part of the `wave run` command, the waved binary is located in the root of the currently active python virtual environment. Local paths need to be relative to this location. Another quick trick to find the waved location is to check `web-dir` during app startup which is located in the same directory as waved itself.

```sh {10}
wave run app --no-reload
2023/09/12 14:24:35 #
2023/09/12 14:24:35 # ┌────────────────┐ H2O Wave
2023/09/12 14:24:35 # │  ┐┌┐┐┌─┐┌ ┌┌─┐ │ 0.26.2 20230803101411
2023/09/12 14:24:35 # │  └┘└┘└─└└─┘└── │ © 2021 H2O.ai, Inc.
2023/09/12 14:24:35 # └────────────────┘
2023/09/12 14:24:35 # ┌──────────────────────────────────────┐
2023/09/12 14:24:35 # │  Running at http://localhost:10101/  │
2023/09/12 14:24:35 # └──────────────────────────────────────┘
2023/09/12 14:24:35 # {"address":":10101","base-url":"/","t":"listen","web-dir":"/path/to/venv/www"}
INFO:     Started server process [10429]
INFO:     Waiting for application startup.
2023/09/12 14:24:36 # {"host":"http://127.0.0.1:8000","route":"/","t":"app_add"}
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
```

From the above, the root of the active virtual python environment is `/path/to/venv`. Another, more straightforward, option is to use absolute paths.

Wave serves whole directories as they are. This means that these directories are listable by default. If you wish to turn off this behavior, simply put an empty file called `index.html` into the folder you wish to not list.

### TLS verification

During development, you might want to test out TLS encryption, e.g. communication between Wave server and Keycloak. The easiest thing to do is to generate a self-signed certificate. However, Wave server verifies certificates for all communication by default, thus would throw an error for a self-signed one. **_FOR DEVELOPMENT PURPOSES ONLY_**, it's possible to turn off the check using either `H2O_WAVE_NO_TLS_VERIFY` environment variable or `no-tls-verify` parameter.

:::warning
**Disabling TLS verification is a security risk.** Make sure TLS is not disabled in production environments.
:::

## Configuring your app

Your Wave application is an ASGI server. When you run your app during development, the app server runs at http://127.0.0.1:8000/ by default (localhost, port 8000), and assumes that your Wave server is running at http://127.0.0.1:10101/ (localhost, port 10101). The `wave run` command automatically picks another available port if `8000` is not available.

The Wave server and apps communicate with each other using RPC over persistent HTTP connections.

For production deployments, you'll want to configure which port your app listens to, how it can access the Wave server, and how the Wave server can access your app.

You can use the following environment variables to configure your app's server's behavior:

### H2O_WAVE_APP_ADDRESS

The public host/port of the app server. Defaults to `http://127.0.0.1:8000`. Set this variable if you are running your Wave server and your app on different machines or containers.

### H2O_WAVE_APP_MODE

The [realtime sync mode](realtime.md) of the app server. One of `unicast` (default), `multicast`, or `broadcast`.

### H2O_WAVE_ADDRESS

The public host/port of the Wave server. Defaults to `http://127.0.0.1:10101`. Set this variable if you are running the Wave server on a remote machine or container. Another common usage for this var is changing the app port together with `H2O_WAVE_LISTEN`.

### H2O_WAVE_CONNECTION_TIMEOUT

The number of seconds to attempt to connect to the Wave server before giving up.

### H2O_WAVE_ACCESS_KEY_ID

The API access key ID to use when communicating with the Wave server.

### H2O_WAVE_ACCESS_KEY_SECRET

The API access key secret to use when communicating with the Wave server.

### H2O_WAVE_APP_ACCESS_KEY_ID

The API access key ID to use when communicating with the app server. Automatically generated if not specfied.

### H2O_WAVE_APP_ACCESS_KEY_SECRET

The API access key secret to use when communicating with the app server. Automatically generated if not specified.

### H2O_WAVE_CHECKPOINT_DIR

The directory to save/load application and session state. Enables checkpointing. If set, the app saves the contents of `q.app` and `q.user` before exiting. When restarted, the contents of `q.app` and `q.user` are restored. The directory is automatically created if it does not exist.

You can use checkpointing as a simple way to save/load your app's data while prototyping.


CHANGELOG.md
# Changelog

Legend:
- `Added` for new features.
- `Changed` for changes in existing functionality.
- `Deprecated` for soon-to-be removed features.
- `Removed` for now removed features.
- `Fixed` for any bug fixes.
- `Security` in case of vulnerabilities.



conf.go
// Copyright 2020 H2O.ai, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package wave

import (
	"net/http"
	"time"

	"github.com/h2oai/wave/pkg/keychain"
)

// ServerConf represents Server configuration options.
type ServerConf struct {
	Version              string
	BuildDate            string
	Listen               string
	BaseURL              string
	WebDir               string
	DataDir              string
	PublicDirs           []string
	PrivateDirs          []string
	Keychain             *keychain.Keychain
	Init                 string
	Compact              string
	CertFile             string
	SkipCertVerification bool
	KeyFile              string
	Header               http.Header
	Editable             bool
	MaxRequestSize       int64
	MaxCacheRequestSize  int64
	Proxy                bool
	MaxProxyRequestSize  int64
	MaxProxyResponseSize int64
	NoStore              bool
	NoLog                bool
	IDE                  bool
	Debug                bool
	Auth                 *AuthConf
	ForwardedHeaders     map[string]bool
	KeepAppLive          bool
	PingInterval         time.Duration
	ReconnectTimeout     time.Duration
	AllowedOrigins       map[string]bool
}

type AuthConf struct {
	ClientID              string
	ClientSecret          string
	ProviderURL           string
	RedirectURL           string
	EndSessionURL         string
	PostLogoutRedirectURL string
	Scopes                []string
	URLParameters         [][]string
	SkipLogin             bool
	SessionExpiry         time.Duration
	InactivityTimeout     time.Duration
}

type Conf struct {
	Version               bool   `cfg:"version" env:"H2O_WAVE_VERSION" cfgDefault:"false"`
	Listen                string `cfg:"listen" env:"H2O_WAVE_LISTEN" cfgDefault:":10101" cfgHelper:"listen on this address"`
	BaseUrl               string `cfg:"base-url" env:"H2O_WAVE_BASE_URL" cfgDefault:"/" cfgHelper:"the base URL (path prefix) to be used for resolving relative URLs (e.g. /foo/ or /foo/bar/, without the host)"`
	WebDir                string `cfg:"web-dir" env:"H2O_WAVE_WEB_DIR" cfgDefault:"./www" cfgHelper:"directory to serve web assets from, hosted at /"`
	DataDir               string `cfg:"data-dir" env:"H2O_WAVE_DATA_DIR" cfgDefault:"./data" cfgHelper:"directory to store site data"`
	PublicDirs            string `cfg:"public-dir" env:"H2O_WAVE_PUBLIC_DIR" cfgDefault:"" cfgHelper:"additional directory to serve files from, in the format \"[url-path]@[filesystem-path]\", e.g. \"/public/files/@/some/local/path\" will host /some/local/path/foo.txt at /public/files/foo.txt; multiple directory mappings allowed"`
	PrivateDirs           string `cfg:"private-dir" env:"H2O_WAVE_PRIVATE_DIR" cfgDefault:"" cfgHelper:"additional directory to serve files from (authenticated users only), in the format \"[url-path]@[filesystem-path]\", e.g. \"/public/files/@/some/local/path\" will host /some/local/path/foo.txt at /public/files/foo.txt; multiple directory mappings allowed"`
	AccessKeyID           string `cfg:"access-key-id" env:"H2O_WAVE_ACCESS_KEY_ID" cfgDefault:"access_key_id" cfgHelper:"default API access key ID"`
	AccessKeySecret       string `cfg:"access-key-secret" env:"H2O_WAVE_ACCESS_KEY_SECRET" cfgDefault:"access_key_secret" cfgHelper:"default API access key secret"`
	AccessKeyFile         string `cfg:"access-keychain" env:"H2O_WAVE_ACCESS_KEYCHAIN" cfgDefault:".wave-keychain" cfgHelper:"path to file containing API access keys"`
	CreateAccessKey       bool   `cfg:"create-access-key" env:"H2O_WAVE_CREATE_ACCESS_KEY" cfgDefault:"false" cfgHelper:"generate and add a new API access key ID and secret pair to the keychain"`
	ListAccessKeys        bool   `cfg:"list-access-keys" env:"H2O_WAVE_LIST_ACCESS_KEYS" cfgDefault:"false" cfgHelper:"list all the access key IDs in the keychain"`
	RemoveAccessKeyID     string `cfg:"remove-access-key" env:"H2O_WAVE_REMOVE_ACCESS_KEY" cfgDefault:"" cfgHelper:"remove the specified API access key ID from the keychain"`
	Init                  string `cfg:"init" env:"H2O_WAVE_INIT" cfgDefault:"" cfgHelper:"initialize site content from AOF log"`
	Compact               string `cfg:"compact" env:"H2O_WAVE_COMPACT" cfgDefault:"" cfgHelper:"compact AOF log"`
	CertFile              string `cfg:"tls-cert-file" env:"H2O_WAVE_TLS_CERT_FILE" cfgDefault:"" cfgHelper:"path to certificate file (TLS only)"`
	KeyFile               string `cfg:"tls-key-file" env:"H2O_WAVE_TLS_KEY_FILE" cfgDefault:"" cfgHelper:"path to private key file (TLS only)"`
	SkipCertVerification  bool   `cfg:"no-tls-verify" env:"H2O_WAVE_NO_TLS_VERIFY" cfgDefault:"false" cfgHelper:"do not verify TLS certificates during external communication - DO NOT USE IN PRODUCTION"`
	HttpHeadersFile       string `cfg:"http-headers-file" env:"H2O_WAVE_HTTP_HEADERS_FILE" cfgDefault:"" cfgHelper:"path to a MIME-formatted file containing additional HTTP headers to add to responses from the server"`
	ForwardedHttpHeaders  string `cfg:"forwarded-http-headers" env:"H2O_WAVE_FORWARDED_HTTP_HEADERS" cfgDefault:"*" cfgHelper:"comma-separated list of case insesitive HTTP header keys to forward to the Wave app from the browser WS connection. If not specified, defaults to '*' - all headers are allowed. If set to an empty string, no headers are forwarded."`
	Editable              bool   `cfg:"editable" env:"H2O_WAVE_EDITABLE" cfgDefault:"false" cfgHelper:"allow users to edit web pages"`
	MaxRequestSize        string `cfg:"max-request-size" env:"H2O_WAVE_MAX_REQUEST_SIZE" cfgDefault:"5M" cfgHelper:"maximum allowed size of HTTP requests to the server (e.g. 5M or 5MB or 5MiB)"`
	MaxCacheRequestSize   string `cfg:"max-cache-request-size" env:"H2O_WAVE_MAX_CACHE_REQUEST_SIZE" cfgDefault:"5M" cfgHelper:"maximum allowed size of HTTP requests to the server cache (e.g. 5M or 5MB or 5MiB)"`
	Proxy                 bool   `cfg:"proxy" env:"H2O_WAVE_PROXY" cfgDefault:"false" cfgHelper:"enable HTTP proxy (for IDE / language server support only - not recommended for internet-facing websites)"`
	MaxProxyRequestSize   string `cfg:"max-proxy-request-size" env:"H2O_WAVE_MAX_PROXY_REQUEST_SIZE" cfgDefault:"5M" cfgHelper:"maximum allowed size of proxied HTTP requests (e.g. 5M or 5MB or 5MiB)"`
	MaxProxyResponseSize  string `cfg:"max-proxy-response-size" env:"H2O_WAVE_MAX_PROXY_RESPONSE_SIZE" cfgDefault:"5M" cfgHelper:"maximum allowed size of proxied HTTP responses (e.g. 5M or 5MB or 5MiB)"`
	SessionExpiry         string `cfg:"session-expiry" env:"H2O_WAVE_SESSION_EXPIRY" cfgDefault:"720h" cfgHelper:"session cookie lifetime duration (e.g. 1800s or 30m or 0.5h)"`
	InactivityTimeout     string `cfg:"session-inactivity-timeout" env:"H2O_WAVE_SESSION_INACTIVITY_TIMEOUT" cfgDefault:"30m" cfgHelper:"session inactivity timeout duration (e.g. 1800s or 30m or 0.5h)"`
	PingInterval          string `cfg:"ping-interval" env:"H2O_WAVE_PING_INTERVAL" cfgDefault:"50s" cfgHelper:"how often should ping messages be sent (e.g. 60s or 1m or 0.1h) to keep the websocket connection alive (default 50s)"`
	NoStore               bool   `cfg:"no-store" env:"H2O_WAVE_NO_STORE" cfgDefault:"false" cfgHelper:"disable storage (scripts and multicast/broadcast apps will not work)"`
	NoLog                 bool   `cfg:"no-log" env:"H2O_WAVE_NO_LOG" cfgDefault:"false" cfgHelper:"disable AOF logging (connect/disconnect and diagnostic logging messages are not disabled)"`
	Debug                 bool   `cfg:"debug" env:"H2O_WAVE_DEBUG" cfgDefault:"false" cfgHelper:"enable debug mode (profiling, inspection, etc.)"`
	ClientID              string `cfg:"oidc-client-id" env:"H2O_WAVE_OIDC_CLIENT_ID" cfgDefault:"" cfgHelper:"OIDC client ID"`
	ClientSecret          string `cfg:"oidc-client-secret" env:"H2O_WAVE_OIDC_CLIENT_SECRET" cfgDefault:"" cfgHelper:"OIDC client secret"`
	ProviderUrl           string `cfg:"oidc-provider-url" env:"H2O_WAVE_OIDC_PROVIDER_URL" cfgDefault:"" cfgHelper:"OIDC provider URL"`
	RedirectUrl           string `cfg:"oidc-redirect-url" env:"H2O_WAVE_OIDC_REDIRECT_URL" cfgDefault:"" cfgHelper:"OIDC redirect URL"`
	EndSessionUrl         string `cfg:"oidc-end-session-url" env:"H2O_WAVE_OIDC_END_SESSION_URL" cfgDefault:"" cfgHelper:"OIDC end session URL"`
	PostLogoutRedirectUrl string `cfg:"oidc-post-logout-redirect-url" env:"H2O_WAVE_OIDC_POST_LOGOUT_REDIRECT_URL" cfgDefault:"" cfgHelper:"OIDC post logout redirect URL"`
	RawAuthScopes         string `cfg:"oidc-scopes" env:"H2O_WAVE_OIDC_SCOPES" cfgDefault:"openid,profile" cfgHelper:"OIDC scopes, comma-separated (default \"openid,profile\")"`
	RawAuthURLParams      string `cfg:"oidc-auth-url-params" env:"H2O_WAVE_OIDC_AUTH_URL_PARAMS" cfgDefault:"" cfgHelper:"additional URL parameters to pass during OIDC authorization, in the format \"key:value\", comma-separated, e.g. \"foo:bar,qux:42\""`
	SkipLogin             bool   `cfg:"oidc-skip-login" env:"H2O_WAVE_OIDC_SKIP_LOGIN" cfgDefault:"false" cfgHelper:"do not display the login form during OIDC authorization"`
	KeepAppLive           bool   `cfg:"keep-app-live" env:"H2O_WAVE_KEEP_APP_LIVE" cfgDefault:"false" cfgHelper:"do not unregister unresponsive apps"`
	Conf                  string `cfg:"conf" env:"H2O_WAVE_CONF" cfgDefault:".env" cfgHelper:"path to configuration file"`
	ReconnectTimeout      string `cfg:"reconnect-timeout" env:"H2O_WAVE_RECONNECT_TIMEOUT" cfgDefault:"5s" cfgHelper:"Time to wait for reconnect before dropping the client"`
	AllowedOrigins        string `cfg:"allowed-origins" env:"H2O_WAVE_ALLOWED_ORIGINS" cfgDefault:"" cfgHelper:"comma-separated list of allowed origins (e.g. http://foo.com) for websocket upgrades, use '*' to allow all origins"`
}
